### PR TITLE
Sudhi staging

### PR DESCRIPTION
**Pull request by elinw**
This pull requests merges work from several branches to improve support for multiple databases. It also passes a checkstyle scan.

### CODE REVIEW

File: libraries/joomla/database/databasequery.php [hunk 1]
```diff
@@ -798,16 +806,18 @@ public function innerJoin($condition)
 	 * $query->insert('#__a)->columns('id, title')->values('1,2')->values->('3,4');
 	 * $query->insert('#__a)->columns('id, title')->values(array('1,2', '3,4'));
 	 *
-	 * @param   mixed  $table  The name of the table to insert data into.
+	 * @param   mixed    $table            The name of the table to insert data into.
+	 * @param   boolean  $increment_field  The name of the field to auto increment.
```
Comment (eddieajau) [inline libraries/joomla/database/databasequery.php]: I missed this one in my review - I don't know the use case of this argument.

**Comment by eddieajau** [comment]
Except for that one query, I think this looks right now.

**Comment by chdemko** [comment]
Going to review this pull request and merging if good for me

File: libraries/joomla/database/database/sqlsrvquery.php [hunk 2]
```diff
@@ -76,7 +76,16 @@ public function __toString()
...
-					$query = 'SET IDENTITY_INSERT ' . $tableName . ' ON;' . $query . 'SET IDENTITY_INSERT ' . $tableName . ' OFF;';
+					if ($this->auto_increment_field)
+					{
+						$query = 'SET IDENTITY_INSERT ' . $tableName . ' ON;' . $query . 'SET IDENTITY_INSERT ' . $tableName . ' OFF;';
+					}
+
+					if ($this->where)
+					{
+						$query .= (string) $this->where;
+					}
```
Comment (chdemko) [inline libraries/joomla/database/database/sqlsrvquery.php:87]: I suppose

```
      if ($this->where)
      {
          $query .= (string) $this->where;
      }
```

must be before the ($this->auto_increment_field) test

File: libraries/joomla/database/databasequery.php (hunk 1 above)
Comment (chdemko) [inline libraries/joomla/database/databasequery.php]: it's used in libraries/joomla/database/database/sqlsrvquery.php  (see file above)

File: libraries/joomla/html/html/rules.php [hunk 3]
```diff
@@ -172,17 +172,19 @@ protected static function _getUserGroups()
...
 		// Get the user groups from the database.
 		$db->setQuery(
-			'SELECT a.id AS value, a.title AS text, COUNT(DISTINCT b.id) AS level' . ' , GROUP_CONCAT(b.id SEPARATOR \',\') AS parents'
-			. ' FROM #__usergroups AS a' . ' LEFT JOIN #__usergroups AS b ON a.lft > b.lft AND a.rgt < b.rgt' . ' GROUP BY a.id'
-			. ' ORDER BY a.lft ASC'
+			'SELECT a.id AS value, a.title AS text, COUNT(DISTINCT b.id) AS level' .
+			' FROM ' . $db->nq('#__usergroups') . ' AS a ' .
+			' LEFT JOIN ' . $db->nq('#__usergroups') . ' AS b ON a.lft > b.lft AND a.rgt < b.rgt' .
+			' GROUP BY a.id,a.id, a.title, a.lft' .
+			' ORDER BY a.lft ASC'
```
Comment (chdemko) [inline libraries/joomla/html/html/rules.php]: Why do you need a.id, a.id (twice) ?

File: libraries/joomla/database/database/sqlsrv.php [hunk 4]
```diff
@@ -921,4 +1014,63 @@ protected function limit($sql, $limit, $offset)
...
+		$this->setQuery("sp_rename '" . $oldTable . "', '" . $newTable . "'");
+
+		return $this->query();
+	}
+
+	/**
+	 * Locks a table in the database.
+	 *
+	 * @param   string  $tableName  The name of the table to unlock.
+	 *
```
Comment (chdemko) [inline libraries/joomla/database/database/sqlsrv.php:1047]: must be to lock ?

File: libraries/joomla/database/database/sqlsrv.php [hunk 5]
```diff
@@ -403,7 +412,7 @@ public function getTableColumns($tables, $typeOnly = true)
 			{
 				foreach ($fields as $field)
 				{
-					$result[$table][$field->Field] = $field;
+					$result[$field->Field] = $field;
 				}
```
Comment (chdemko) [inline libraries/joomla/database/database/sqlsrv.php]: Why did you remove [$table] ?

File: libraries/joomla/database/database/sqlsrv.php (hunk 4 above)
Comment (elinw) [inline libraries/joomla/database/database/sqlsrv.php:1047]: https://github.com/eddieajau/joomla-platform/blob/sudhi-staging/libraries/joomla/database/database/sqlsrv.php#L1052

Looks like a possible copy and paste issue.

Comment (eddieajau) [inline libraries/joomla/database/database/sqlsrv.php:1047]: s/unlock/lock/

File: libraries/joomla/database/databasequery.php (hunk 1 above)
Comment (eddieajau) [inline libraries/joomla/database/databasequery.php]: I get that, but I'm not sure how it can be used in practice.  Either way, would be nicer to have the variable in camel case "autoIncrementField"

File: libraries/joomla/html/html/rules.php (hunk 3 above)
Comment (eddieajau) [inline libraries/joomla/html/html/rules.php]: You don't - it's a typo that someone, probably me, missed.

File: libraries/joomla/database/database/sqlsrv.php (hunk 5 above)
Comment (eddieajau) [inline libraries/joomla/database/database/sqlsrv.php]: Because that's how it should be.  See:
https://github.com/joomla/joomla-platform/blob/staging/libraries/joomla/database/database/mysql.php#L378

**Comment by eddieajau** [comment]
Thanks for the extra set of eyes on this Christophe.

File: libraries/joomla/database/database/sqlsrv.php (hunk 5 above)
Comment (chdemko) [inline libraries/joomla/database/database/sqlsrv.php]: So I suppose there is an issue at line 407

```
$result[$table][$field->Field] = preg_replace("/[(0-9)]/", '', $field->Type);
```

should be

```
$result[$field->Field] = preg_replace("/[(0-9)]/", '', $field->Type);
```

File: libraries/joomla/html/html/rules.php [hunk 6]
```diff
@@ -172,17 +172,19 @@ protected static function _getUserGroups()
 
 		// Get the user groups from the database.
 		$db->setQuery(
-			'SELECT a.id AS value, a.title AS text, COUNT(DISTINCT b.id) AS level' . ' , GROUP_CONCAT(b.id SEPARATOR \',\') AS parents'
-			. ' FROM #__usergroups AS a' . ' LEFT JOIN #__usergroups AS b ON a.lft > b.lft AND a.rgt < b.rgt' . ' GROUP BY a.id'
```
Comment (chdemko) [inline libraries/joomla/html/html/rules.php:176]: The <code>parents</code> is not selected in the new code but it's used below

File: libraries/joomla/html/html/rules.php [hunk 7]
```diff
@@ -172,17 +172,19 @@ protected static function _getUserGroups()
...
+			' ORDER BY a.lft ASC'
 		);
 		$options = $db->loadObjectList();
 
 		// Pre-compute additional values.
 		foreach ($options as &$option)
 		{
-			// Pad the option text with spaces using depth level as a multiplier.
 
+			$option->parents = '';
```
Comment (chdemko) [inline libraries/joomla/html/html/rules.php]: This should be the result of the select ?

File: libraries/joomla/html/html/rules.php (hunk 3 above)
Comment (hooduku) [inline libraries/joomla/html/html/rules.php]: We can remove the extra a.id.

File: libraries/joomla/database/database/sqlsrv.php (hunk 5 above)
Comment (eddieajau) [inline libraries/joomla/database/database/sqlsrv.php]: Yes, line 407 is wrong too.
